File: src/main/assembly/help/Tutorial_py/4_AdvancedScans/41_ManualScan.py
################################################################################################### 
#Manual scan: Manually setting positioners and reading back sensors, but still using 
#the standard data handling and plotting of built-in scans.
################################################################################################### 


MOTOR_RANGE = (0.0, 8.0)
OUTPUT_SETPOINTS = (1.0, 2.0, 3.0)
FIXED_X = True


writables_names = [m1.getName(), ao1.getName()]
readable_names = [ai1.getName(), ai2.getName()]
start =  [ MOTOR_RANGE[0] if FIXED_X else -1, OUTPUT_SETPOINTS[0]] 
stop = [ MOTOR_RANGE[1] if FIXED_X else -1, OUTPUT_SETPOINTS[-1]]
steps = [int(MOTOR_RANGE[1]-MOTOR_RANGE[0]), len(OUTPUT_SETPOINTS)-1]

scan = ManualScan(writables_names, readable_names ,start, stop, steps)


#This option is to plot the foe each output value one 1D series, instead of all in a matrix plot
setup_plotting(line_plots = (ai1, ai2))


scan.start()
m1.setSpeed(10.0)
for setpoint1 in frange(MOTOR_RANGE[0], MOTOR_RANGE[1], 1.0, True):
    m1.move(setpoint1)
    for setpoint2 in OUTPUT_SETPOINTS:
        ao1.write(setpoint2)
        scan.append ([setpoint1, setpoint2], [m1.read(), ao1.read()], [ai1.read(), ai2.read()])


scan.end()

File: pshell-workbench/src/main/assembly/help/Tutorial_js/4_AdvancedScans/41_ManualScan.js
///////////////////////////////////////////////////////////////////////////////////////////////////
// Manual scan: Manually setting positioners and reading back sensors, but still using 
// the standard data handling and plotting of built-in scans.
///////////////////////////////////////////////////////////////////////////////////////////////////


var MOTOR_RANGE = [0.0, 8.0]
var OUTPUT_SETPOINTS = [1.0, 2.0, 3.0]
var FIXED_X = true


writables_names = to_array([m1.getName(), ao1.getName()], 's')
readable_names = to_array([ai1.getName(), ai2.getName()], 's')
start = to_array( [ FIXED_X ? MOTOR_RANGE[0] : -1, OUTPUT_SETPOINTS[0]] , 'd')
stop =to_array([ FIXED_X ? MOTOR_RANGE[1] : -1, OUTPUT_SETPOINTS[OUTPUT_SETPOINTS.length-1]] , 'd')
steps = to_array([Math.round(MOTOR_RANGE[1]-MOTOR_RANGE[0]), OUTPUT_SETPOINTS.length-1] , 'i')

print (to_array(writables_names))
print (to_array(readable_names))
print (to_array(start))
print (to_array(stop))
print (to_array(steps))

scan = new ManualScan(writables_names, readable_names, start, stop,  steps, false)


//This option is to plot the foe each output value one 1D series, instead of all in a matrix plot
set_preference(Preference.PLOT_TYPES, {"ai1":1, "ai2":1})


scan.start()
m1.setSpeed(10.0)
for (var setpoint1  = MOTOR_RANGE[0]; setpoint1 <=MOTOR_RANGE[1]; setpoint1+=1.0){
    m1.move(setpoint1)
    for (var setpoint2 in OUTPUT_SETPOINTS){
        ao1.write(setpoint2)
        scan.append (to_array([setpoint1, setpoint2], 'd'), to_array([m1.read(), ao1.read()], 'd'), to_array([ai1.read(), ai2.read()], 'd'))
    }
}

scan.end()
